src/Compilers/Core/Portable/Compilation/Compilation.cs

```diff
@@ -1372,14 +1376,19 @@ internal bool FilterAndAppendDiagnostics(DiagnosticBag accumulator, IEnumerable<
                     continue;
                 }
 
-                var filtered = Options.FilterDiagnostic(d);
-                if (filtered == null ||
-                    (!reportSuppressedDiagnostics && filtered.IsSuppressed))
+                Diagnostic filtered = d;
+                if (filterDiagnostics)
                 {
-                    continue;
+                    filtered = Options.FilterDiagnostic(d);
+                    if (filtered == null ||
+                        (!reportSuppressedDiagnostics && filtered.IsSuppressed))
+                    {
+                        continue;
+                    }
                 }
-                else if (filtered.Severity == DiagnosticSeverity.Error &&
-                         !filtered.IsSuppressed)
+
+                if (filtered.Severity == DiagnosticSeverity.Error &&
+                    !filtered.IsSuppressed)
                 {
                     hasError = true;
```

Comment:
>hasError = true; [](start = 20, length = 16)

It looks like we might fail to get here because ```Options.FilterDiagnostic(d)``` can be skipped above.

Reply:
`Options.FilterDiagnostic(d)` will be skipped only if we have explicitly requested to skip filtering. If so, we will apply the filtering at a [later stage](https://github.com/dotnet/roslyn/pull/36067/files/08538a823676aeff914b1f61e978ebf6094f037a#diff-a969f68cd2bc892509eb83bec00d060eR1067).

Reply:
> I thought this would be useful for the callers, I can remove this comment if it seems to add more confusion.

That is likely to cause an observable behavior difference. Why would we want that?

---
In reply to: [289971344](https://github.com/dotnet/roslyn/pull/36067#discussion_r289971344) [](ancestors = 289971344)

Reply:
Reverted